### PR TITLE
add minimalvec

### PR DESCRIPTION
**Pull request by Jutho**

### CODE REVIEW

**Comment by Jutho** [comment]
Ok I think this is ready. Maybe some review is welcome, @lkdvos ?

File: test/minimalmvec.jl
Comment (lkdvos) [inline test/minimalmvec.jl:1]: In principle, this is quite an extensive testset for something that is simply wrapping `Vector`. We could simply have:

```julia
@testset "scale" begin
    a = rand(2)
    alfa = rand()
    @test scale(a, alfa) == scale(MinimalMVec(a), alfa).vec 
    ...
end
```
Along with some simple checks of type stability. I am also fine with just leaving this as is, but less code is easier to maintain 😉

File: test/minimalmvec.jl [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+deepcollect(x::Number) = x
+
+x = MinimalMVec(randn(3))
+y = MinimalMVec(randn(3))
+
+@testset "scalartype" begin
+    s = @constinferred scalartype(x)
+    @test s == Float64
+    @test_throws ArgumentError scalartype(())
+    @test_throws ArgumentError scalartype((randn(Float64, 2), randn(ComplexF64, 3)))
```
Comment (lkdvos) [inline test/minimalmvec.jl]:
```suggestion
```

File: test/minimalsvec.jl [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+deepcollect(x::Number) = x
+
+x = MinimalSVec(randn(3))
+y = MinimalSVec(randn(3))
+
+@testset "scalartype" begin
+    s = @constinferred scalartype(x)
+    @test s == Float64
+    @test_throws ArgumentError scalartype(())
+    @test_throws ArgumentError scalartype((randn(Float64, 2), randn(ComplexF64, 3)))
```
Comment (lkdvos) [inline test/minimalsvec.jl]:
```suggestion
```

**Review by lkdvos — APPROVED** [review]

File: test/minimalmvec.jl
Comment (Jutho) [reply test/minimalmvec.jl:1]: That is true, it's copy pasted code, with small adjustments for the specific case. 

It did help me with finding some problem in our fallbacks though. Base has a `mul!(::Any, :Any, :Any)` method, so testing `applicable(mul!, x, y, alpha)` was kind of pointless. That `Tuple{Any,Any,Any}` method assumes this is a matrix multiplication call, and just adds 2 more arguments `true` and `false` playing the role of alpha and beta. But that is of course pointless if you want to do what we call `scale!` or `add!`, and `alpha` is already the third argument. I guess this just illustrates what I dislike about the current `mul!` method and why I started VectorInterface.jl in the first place. 

It actually also helped me track down another error in our `MinimalVec` implementation, where our interface promises that `scale!!(y, x, alpha)` will always work. So even if `y` is mutable, but happens to have a scalar type that is not compatible with the scalar type of x * alpha, the method should not complain and return a new vector (which actually has scalar type that is the promotion of that of y, x and alpha)`. This was broken and resulting in an error in the original implementation.

So I would favor keeping it for now.

**Pull request merged**